### PR TITLE
picolibc: Update to version 1.8.7

### PR DESCRIPTION
**Pull request by keith-packard**
Integrate new upstream release

The associated sdk-ng work is happening on https://github.com/zephyrproject-rtos/sdk-ng/tree/topic-picolibc-1.8.7

### CODE REVIEW

File: lib/libc/common/source/time/ctime.c [hunk 1]
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#undef _POSIX_C_SOURCE
```
Comment (cfriedt) [inline lib/libc/common/source/time/ctime.c:7]: I understand why the kludge is necessary. I was hoping to address this particular issue preemptively almost 9 months ago :'(

https://github.com/zephyrproject-rtos/zephyr/pull/67132

Comment (cfriedt) [reply lib/libc/common/source/time/ctime.c:7]: I'm happy to approve as-is, but @keith-packard, if you feel like we could use the Kconfig, I would be more than happy to pursue that PR again.

Comment (keith-packard) [reply lib/libc/common/source/time/ctime.c:7]: Up to you; we can revert this change if that happens.

File: subsys/net/lib/dns/mdns_responder.c [hunk 2]
```diff
@@ -35,6 +35,18 @@ LOG_MODULE_REGISTER(net_mdns_responder, CONFIG_MDNS_RESPONDER_LOG_LEVEL);
 
 #include "net_private.h"
 
+/*
+ * GCC complains about struct sockaddr accesses due to the various
+ * address-family-specific variants being of differing sizes. Let's not
+ * mess with code (which looks correct), just silence the compiler.
```
Comment (jukkar) [inline subsys/net/lib/dns/mdns_responder.c:41]: Interesting. I have not seen these issues so far, what gcc version is this?

Also could you add `net: ` prefix to the dns commit subject, it is just easier to find it later if needed.
So `net: dns: ....`

**Review by aescolar — DISMISSED** [review]

File: subsys/net/lib/dns/mdns_responder.c (hunk 2 above)
Comment (keith-packard) [reply subsys/net/lib/dns/mdns_responder.c:41]: Looks like I'm using gcc 14.2.0 on Debian unstable. I'lve fixed the dns: commit message, now to go fix more build failures.

**Review by jukkar — DISMISSED** [review]

**Comment by keith-packard** [comment]
Given the number of changes in the POSIX code, I'd love to get @cfriedt to give this another glance before it gets merged?

File: lib/libc/common/Kconfig [hunk 3]
```diff
@@ -8,26 +8,30 @@ config COMMON_LIBC_ABORT
 
 config COMMON_LIBC_ASCTIME
 	bool
+	depends on !PICOLIBC
 	help
 	  common implementation of asctime().
 
 config COMMON_LIBC_ASCTIME_R
 	bool "Thread-safe version of asctime()"
 	default y if POSIX_THREAD_SAFE_FUNCTIONS
```
Comment (cfriedt) [inline lib/libc/common/Kconfig:17]: These lines should be removed, and instead in a POSIX Kconfig, we should use a pattern like:

```
config POSIX_C_LANG_SUPPORT_R
    select LIBC_COMMON_ASCTIME_R if ! TC_PROVIDES_POSIX_C_LANG_SUPPORT_R
    ...
```

That way picolibc could just select `TC_PROVIDES_POSIX_C_LANG_SUPPORT_R` and cover all of those functions at once. That might be closer in intent to the way that the TC_PROVIDES macros gate files being built in cmakelists.txt and we don't need to use `depends on ! PICOLIBC`

Comment (keith-packard) [reply lib/libc/common/Kconfig:17]: Had to remove a couple of redundant 'default y if POSIX_THREAD_SAFE_FUNCTIONS' clauses. Looks like a cleaner plan, thanks!

**Review by cfriedt — APPROVED** [review]
LGTM.

I'll respin #67132 as well, so that we can have a way to avoid peppering `_POSIX_C_SOURCE` throughout the code, for consistency.

**Review by ycsin — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**